# **🎬 Project Introduction**

**Title: Advanced Text Generation using LSTM**

"Welcome, everyone! Today, we are taking a massive leap in Deep Learning by exploring one of the most powerful concepts in NLP: **Text Generation**.

If you've worked with Simple RNNs before, you already know they struggle to remember long-term context. Today, we are going to solve that problem practically. We will build a highly capable **LSTM (Long Short-Term Memory) Model** from scratch to generate human-like text. We will see exactly how the internal gates of an LSTM allow it to remember crucial information over long sequences, making it the perfect tool for writing stories, poems, or code. Let’s dive in!"

---

### 🎬 প্রজেক্ট ইন্ট্রোডাকশন (আপডেটেড বাংলা ভার্সন)

**শিরোনাম: LSTM-এর সাহায্যে অ্যাডভান্সড টেক্সট জেনারেশন**

"সবাইকে স্বাগতম! আজকে আমরা ডিপ লার্নিং এবং এনএলপি (NLP)-এর অত্যন্ত চমৎকার একটি প্রজেক্ট নিয়ে কাজ করব— আর তা হলো **Text Generation** বা মেশিনকে দিয়ে নতুন টেক্সট লেখানো।

আপনারা যারা আগে Simple RNN নিয়ে কাজ করেছেন, তারা জানেন যে বড় টেক্সটের ক্ষেত্রে পুরনো কনটেক্সট বা মেমরি ধরে রাখতে RNN কতটা হিমশিম খায়। আজ আমরা প্র্যাকটিক্যালি এই সমস্যার সমাধান করব। আমরা একটি শক্তিশালী **LSTM মডেল** তৈরি করব এবং দেখব কীভাবে এর ভেতরের মেমরি গেটগুলো লম্বা সময়ের জন্য তথ্য মনে রাখতে পারে। গল্প, কবিতা বা যেকোনো হিউম্যান-লাইক টেক্সট জেনারেট করার ক্ষেত্রে এটি কীভাবে জাদুর মতো কাজ করে, চলুন সেটিই আজ কোড করে দেখা যাক!"


# **📖 Language Modeling Concept (English Version)**

Language modeling is the core concept behind text generation. It is the exact mechanism that advanced AI models like ChatGPT use to predict the next word in a sequence to generate human-like text.

Let's look at an example sentence: **"The Last Train Journey"**

If we want to build a model to generate this text, we must frame it as a **Supervised Learning** problem. But how do we define the Features (input) and the Target (output)? We do this by using the previous words to predict the next word:

| Features (X) | Target (y) |
| --- | --- |
| The | Last |
| The Last | Train |
| The Last Train | Journey |

By structuring the data this way, we successfully separate our features and targets. This next-word prediction process is the fundamental concept of Language Modeling.

### 🛠️ Project Steps

To build our text generation project, we will follow these 4 steps:

1. **Form Vocabulary:** Create a dictionary of all the unique words present in our dataset.
2. **Convert Text to Indices:** Since machines only understand numbers, we will convert each word into a unique numerical index.
3. **Form Sequences:** We will structure our features and targets using these numbers.
| Features (Indices) | Target (Index) |
| --- | --- |
| The [1] | Last [2] |
| The [1], Last [2] | Train [3] |
| The [1], Last [2], Train [3] | Journey [4] |


4. **Word Embedding:** Finally, we will convert each numerical index into a dense vector (e.g., a 50-dimensional vector). This allows the machine to understand the semantic meaning and relationships between words. Common techniques discussed in NLP include Word2Vec, GloVe, TF-IDF, or standard Neural Network Embedding layers.

---

### 📖 ল্যাঙ্গুয়েজ মডেলিং কনসেপ্ট (বাংলা ভার্সন)

টেক্সট জেনারেশনের পেছনের মূল ভিত্তি হলো ল্যাঙ্গুয়েজ মডেলিং। ChatGPT-এর মতো অ্যাডভান্সড এআই মডেলগুলো ঠিক এই মেকানিজম ব্যবহার করেই সিকোয়েন্সের পরবর্তী শব্দটি প্রেডিক্ট করে এবং মানুষের মতো টেক্সট তৈরি করে।

ধরা যাক, আমাদের কাছে একটি বাক্য আছে: **"The Last Train Journey"**

আমরা যদি এই টেক্সটটি জেনারেট করার জন্য একটি মডেল তৈরি করতে চাই, তবে এটিকে একটি **Supervised Learning** প্রবলেমে রূপান্তর করতে হবে। কিন্তু এখানে ফিচার (ইনপুট) এবং টার্গেট (আউটপুট) কী হবে? আমরা মূলত আগের শব্দগুলোকে ব্যবহার করে ঠিক তার পরের শব্দটি প্রেডিক্ট করব:

| Features (X) | Target (y) |
| --- | --- |
| The | Last |
| The Last | Train |
| The Last Train | Journey |

এভাবে ডেটা সাজানোর মাধ্যমে আমরা খুব সহজেই ফিচার এবং টার্গেট আলাদা করে ফেলতে পারি। এই "Next Word Prediction" বা পরবর্তী শব্দ অনুমান করার প্রক্রিয়াটিই হলো ল্যাঙ্গুয়েজ মডেলিংয়ের মূল কনসেপ্ট।

### 🛠️ প্রজেক্টের ধাপসমূহ (Project Steps)

আমাদের টেক্সট জেনারেশন প্রজেক্টটি তৈরি করার জন্য আমরা নিচের ৪টি ধাপ অনুসরণ করব:

১. **Vocabulary তৈরি করা:** আমাদের ডেটাসেটে থাকা সব ইউনিক বা আলাদা শব্দগুলোর একটি ডিকশনারি তৈরি করা।
২. **Text to Indices কনভার্সন:** যেহেতু মেশিন সরাসরি টেক্সট বোঝে না, তাই আমরা প্রতিটি শব্দকে একটি ইউনিক নাম্বারে (Index) কনভার্ট করব।
৩. **Sequence তৈরি করা:** নাম্বারগুলো ব্যবহার করে আমরা আমাদের ফিচার এবং টার্গেটের সিকোয়েন্স সাজাব।

| Features (Indices) | Target (Index) |
| --- | --- |
| The [1] | Last [2] |
| The [1], Last [2] | Train [3] |
| The [1], Last [2], Train [3] | Journey [4] |

৪. **Word Embedding:** সবশেষে, প্রতিটি ইন্ডেক্স বা শব্দকে আমরা একটি নির্দিষ্ট ডাইমেনশনের (যেমন: ৫০ ডাইমেনশন) ভেক্টরে রূপান্তর করব। এটি মেশিনকে শব্দের অর্থ এবং একটি শব্দের সাথে অন্য শব্দের সম্পর্ক বুঝতে সাহায্য করে। NLP-তে সাধারণত Word2Vec, GloVe, TF-IDF বা স্ট্যান্ডার্ড এমবেডিং লেয়ার ব্যবহার করে এই কাজটি করা হয়।

# **Import Library and Tokenize dataset**

### 📝 Code Explanation (English Version)

**1. Installing and Importing Libraries**

* `!pip install nltk`: This command installs the Natural Language Toolkit (NLTK) library, which is essential for text processing.
* `import torch`: Imports the main PyTorch library, which we will use to build our LSTM model.
* `import torch.nn as nn`: Imports the neural network modules from PyTorch (used to create layers like Embedding, LSTM, Linear).
* `import torch.optim as optim`: Imports optimization algorithms (like Adam or SGD) to update model weights during training.
* `import numpy as np`: Imports NumPy for efficient numerical and array operations.
* `from collections import Counter`: Imports the `Counter` class. We will use this later to count how many times each word appears in the document to build our vocabulary.
* `from torch.utils.data import Dataset, DataLoader`: Imports PyTorch tools to organize our data into batches and feed it efficiently to the model during training.
* `from nltk.tokenize import word_tokenize`: Imports the specific function from NLTK that splits a long string of text into a list of individual words and punctuation marks.
* `import nltk`: Imports the main NLTK package to access its download modules.

**2. The Dataset**

* `document = """..."""`: This triple-quoted string holds your entire story ("The Last Train Journey"). In NLP, we refer to this raw text data as our **corpus**.

**3. Downloading Tokenizer Models**

* `nltk.download('punkt')`: NLTK needs a set of pre-trained rules to know exactly how to split English sentences (e.g., separating periods from words). This downloads the required 'punkt' tokenizer model.
* `nltk.download('punkt_tab')`: In newer versions of NLTK, this downloads additional mapping tables required by the 'punkt' tokenizer to function correctly.

**4. Tokenization Step**

* `tokens = word_tokenize(document.lower())`: This is the core step!
* First, `document.lower()` converts the entire story into lowercase letters. This ensures that "The" and "the" are treated as the exact same word by our model.
* Then, `word_tokenize(...)` splits the lowercase text into individual words and punctuation marks.


* `tokens`: This outputs the final list of tokens. (e.g., `['the', 'last', 'train', 'journey', ',', 'on', 'a', ...]`).

---

### 📝 কোড এক্সপ্ল্যানেশন (বাংলা ভার্সন)

**১. লাইব্রেরি ইনস্টল এবং ইমপোর্ট করা**

* `!pip install nltk`: এই কমান্ডটির মাধ্যমে NLTK (Natural Language Toolkit) লাইব্রেরি ইনস্টল করা হয়, যা টেক্সট বা ভাষা প্রসেসিংয়ের জন্য অত্যন্ত প্রয়োজনীয়।
* `import torch`: মূল PyTorch লাইব্রেরি ইমপোর্ট করা হলো, যা দিয়ে আমরা আমাদের LSTM মডেল তৈরি করব।
* `import torch.nn as nn`: PyTorch-এর নিউরাল নেটওয়ার্ক মডিউল ইমপোর্ট করা হলো (এটি দিয়ে Embedding, LSTM, Linear লেয়ার তৈরি করা হবে)।
* `import torch.optim as optim`: অপটিমাইজেশন অ্যালগরিদম (যেমন: Adam বা SGD) ইমপোর্ট করা হলো, যা ট্রেনিংয়ের সময় মডেলের ওয়েট আপডেট করবে।
* `import numpy as np`: নিউমেরিক্যাল বা ম্যাথমেটিক্যাল হিসাব-নিকাশ দ্রুত করার জন্য NumPy ইমপোর্ট করা হলো।
* `from collections import Counter`: `Counter` ক্লাসটি ইমপোর্ট করা হলো। আমাদের ডকুমেন্টে কোন শব্দ কতবার এসেছে তা গুনে (frequency) Vocabulary তৈরি করতে এটি পরে কাজে লাগবে।
* `from torch.utils.data import Dataset, DataLoader`: PyTorch-এর এই টুলগুলো ডেটাকে ব্যাচ (Batch) আকারে সাজাতে এবং ট্রেনিংয়ের সময় মডেলের কাছে খুব সুন্দরভাবে ডেটা পাঠাতে সাহায্য করে।
* `from nltk.tokenize import word_tokenize`: NLTK থেকে নির্দিষ্ট ফাংশনটি ইমপোর্ট করা হলো, যার কাজ হলো একটি বড় টেক্সটকে ভেঙে আলাদা আলাদা শব্দে (Tokens) রূপান্তর করা।
* `import nltk`: NLTK-এর মূল প্যাকেজ ইমপোর্ট করা হলো যাতে আমরা এর ডাউনলোড মডিউলগুলো ব্যবহার করতে পারি।

**২. ডেটাসেট (The Dataset)**

* `document = """..."""`: এই ভেরিয়েবলের ভেতর তোমার পুরো গল্পটি ("The Last Train Journey") রাখা হয়েছে। NLP-এর ভাষায় এই মূল টেক্সট ডেটাকে বলা হয় **Corpus (কর্পাস)**।

**৩. টোকেনাইজার মডেল ডাউনলোড করা**

* `nltk.download('punkt')`: ইংরেজি বাক্যগুলোকে ঠিক কীভাবে ভাঙতে হবে (যেমন: শব্দের সাথে থাকা ফুলস্টপকে আলাদা করা), তার জন্য NLTK-এর কিছু আগে থেকে তৈরি করা রুলস বা মডেল লাগে। এই কমান্ডটি সেই 'punkt' টোকেনাইজার মডেলটি ডাউনলোড করে।
* `nltk.download('punkt_tab')`: NLTK-এর নতুন ভার্সনগুলোতে 'punkt' টোকেনাইজারটি ঠিকমতো কাজ করার জন্য কিছু অতিরিক্ত টেবিল লাগে, এটি মূলত সেই ফাইলগুলো ডাউনলোড করে।

**৪. টোকেনাইজেশন ধাপ (Tokenization Step)**

* `tokens = word_tokenize(document.lower())`: এটিই হলো সবচেয়ে গুরুত্বপূর্ণ কাজ!
* প্রথমে, `document.lower()` ফাংশনটি পুরো গল্পের সমস্ত বড় হাতের অক্ষরকে ছোট হাতের (lowercase) অক্ষরে রূপান্তর করে। এর ফলে মডেল "The" এবং "the"-কে দুটি আলাদা শব্দ না ভেবে একই শব্দ হিসেবে কাউন্ট করবে।
* এরপর, `word_tokenize(...)` সেই ছোট হাতের টেক্সটটিকে ভেঙে প্রতিটি শব্দ এবং যতিচিহ্নকে আলাদা আলাদা উপাদান হিসেবে একটি লিস্টে পরিণত করে।


* `tokens`: এটি লেখার ফলে ফাইনাল টোকেনগুলোর লিস্ট প্রিন্ট হবে (যেমন: `['the', 'last', 'train', 'journey', ',', 'on', 'a', ...]`)।

## **Import Library and Tokenize Dataset**

In [1]:
!pip install nltk

In [2]:
import torch
import torch.nn as nn
import torch.optim as optim
import numpy as np
from collections import Counter #Counts frequency of elements in a list or text.
from torch.utils.data import Dataset, DataLoader
from nltk.tokenize import word_tokenize  #Splits text into individual words/tokens.
import nltk # Used for NLP tasks like tokenization, stemming, stopword removal, etc

In [3]:
document = """The Last Train Journey

On a cold winter evening, Daniel arrived at the old railway station carrying a small leather bag.
The station was nearly empty except for a few passengers waiting quietly on wooden benches.
A loud whistle echoed through the air as the final train of the night slowly entered the platform.

Daniel checked the ticket in his pocket and walked toward the train.
The metal doors opened with a creaking sound and warm air rushed outside.
Inside the train, yellow lights flickered softly above rows of empty seats.

He chose a seat beside the window and placed his bag carefully near his feet.
As the train began to move, the city lights slowly disappeared behind thick clouds of fog.
Snow started falling gently outside while distant mountains appeared under the moonlight.

Across the aisle sat an elderly woman reading an old book with a blue cover.
Every few minutes she looked outside the window as if searching for something familiar.
Daniel noticed a silver necklace around her neck shaped like a small compass.

After several hours, the train stopped at a remote station surrounded by tall pine trees.
Very few people entered or left the train at that place.
A strange silence covered the station as the doors closed once again.

Curious about the woman, Daniel finally started a conversation.
The woman introduced herself as Eleanor and explained that she was returning to her hometown after many years.
She said the town was hidden deep within the northern mountains and was known for its beautiful frozen lake.

As the journey continued, Eleanor shared stories from her childhood.
She described colorful festivals, music performances, and lanterns floating across the lake during winter nights.
Daniel listened carefully while the train moved through snowy valleys and dark forests.

Near midnight, the train suddenly slowed down because heavy snow had covered the tracks ahead.
Passengers became nervous and began discussing possible delays.
Conductors moved through the cabins trying to calm everyone.

Daniel looked outside and saw workers removing snow under bright floodlights.
The freezing wind shook the train slightly while snow continued falling from the dark sky.
Despite the delay, Eleanor remained calm and continued telling her stories.

Several hours later, the tracks were finally cleared and the train resumed its journey.
The passengers sighed with relief as warm coffee was served throughout the cabins.
Daniel and Eleanor continued talking about travel, history, and forgotten places around the world.

Just before sunrise, the train reached the mountain town.
The station was small but beautifully decorated with glowing lamps and snow-covered rooftops.
People wearing thick winter coats walked slowly across the icy streets.

Before leaving, Eleanor handed Daniel a folded piece of paper.
Inside it was a drawing of the frozen lake along with a short message thanking him for the conversation.
Daniel smiled and carefully placed the paper inside his notebook.

As the train departed once again, Daniel stood on the platform watching the sunrise over the mountains.
The cold air, quiet streets, and distant church bells created a peaceful atmosphere he would never forget.
Years later, Daniel would still remember that mysterious winter journey and the stories shared by Eleanor on the last train of the night.
"""

In [4]:
# NLTK does NOT come with pretrained tokenizers by default.

# Tokenization
nltk.download('punkt')

# Some newer NLTK versions require additional tokenizer tables.
nltk.download('punkt_tab')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt_tab.zip.


True

### **Tokenize**

In [5]:
tokens = word_tokenize(document.lower())
tokens

['the',
 'last',
 'train',
 'journey',
 'on',
 'a',
 'cold',
 'winter',
 'evening',
 ',',
 'daniel',
 'arrived',
 'at',
 'the',
 'old',
 'railway',
 'station',
 'carrying',
 'a',
 'small',
 'leather',
 'bag',
 '.',
 'the',
 'station',
 'was',
 'nearly',
 'empty',
 'except',
 'for',
 'a',
 'few',
 'passengers',
 'waiting',
 'quietly',
 'on',
 'wooden',
 'benches',
 '.',
 'a',
 'loud',
 'whistle',
 'echoed',
 'through',
 'the',
 'air',
 'as',
 'the',
 'final',
 'train',
 'of',
 'the',
 'night',
 'slowly',
 'entered',
 'the',
 'platform',
 '.',
 'daniel',
 'checked',
 'the',
 'ticket',
 'in',
 'his',
 'pocket',
 'and',
 'walked',
 'toward',
 'the',
 'train',
 '.',
 'the',
 'metal',
 'doors',
 'opened',
 'with',
 'a',
 'creaking',
 'sound',
 'and',
 'warm',
 'air',
 'rushed',
 'outside',
 '.',
 'inside',
 'the',
 'train',
 ',',
 'yellow',
 'lights',
 'flickered',
 'softly',
 'above',
 'rows',
 'of',
 'empty',
 'seats',
 '.',
 'he',
 'chose',
 'a',
 'seat',
 'beside',
 'the',
 'window',
 'a

In [6]:
Counter(tokens)

Counter({'the': 55,
         'last': 2,
         'train': 14,
         'journey': 4,
         'on': 4,
         'a': 16,
         'cold': 2,
         'winter': 4,
         'evening': 1,
         ',': 19,
         'daniel': 11,
         'arrived': 1,
         'at': 3,
         'old': 2,
         'railway': 1,
         'station': 5,
         'carrying': 1,
         'small': 3,
         'leather': 1,
         'bag': 2,
         '.': 39,
         'was': 7,
         'nearly': 1,
         'empty': 2,
         'except': 1,
         'for': 4,
         'few': 3,
         'passengers': 3,
         'waiting': 1,
         'quietly': 1,
         'wooden': 1,
         'benches': 1,
         'loud': 1,
         'whistle': 1,
         'echoed': 1,
         'through': 3,
         'air': 3,
         'as': 8,
         'final': 1,
         'of': 6,
         'night': 2,
         'slowly': 3,
         'entered': 2,
         'platform': 2,
         'checked': 1,
         'ticket': 1,
         'in': 1,
      

# **Form vocabulary**

### 📖 Why do we build a Vocabulary?
Deep learning models (like LSTMs) are essentially massive calculators. They cannot read or understand raw text (like "train" or "journey"). They only understand numbers (matrices and tensors).

Therefore, we need to create a **"Dictionary" or "Lookup Table"** that maps every unique word in our dataset to a unique integer (number). This mapping is called a Vocabulary.

**Why the `<unk>` token?**
When our model is deployed in the real world, a user might input a word that the model has never seen during training (e.g., "helicopter"). If the model doesn't have an ID for that word, it will crash. To prevent this, we create an `<unk>` (Unknown) token. Any word not found in our vocabulary is automatically assigned the ID of `<unk>` (which is 0), allowing the model to process it without crashing.

### 📝 Line-by-Line Code Explanation

```python
# Initialize the vocabulary dictionary with the unknown token.
vocab = { '<unk>' : 0 }

```

* **What it does:** Creates a Python dictionary named `vocab`. It starts with one entry: the special token `'<unk>'`, which is assigned the ID/index `0`.

```python
# Loop through all the unique words in our tokenized text.
for token in Counter(tokens).keys():

```

* **What it does:** `Counter(tokens)` calculates how many times every word appears. By adding `.keys()`, we ignore the frequency counts and only extract a list of all the **unique** words. The `for` loop goes through this list of unique words one by one.

```python
# Check if the word is already in the dictionary.
  if token not in vocab:

```

* **What it does:** This is a safety check. It ensures we don't accidentally overwrite an existing token (like if `'<unk>'` happened to be in the text).

```python
# Assign a new, unique numerical ID to the word.
    vocab[token] = len(vocab)

```

* **What it does:** This is the most clever part of the code! It assigns the *current length* of the dictionary as the new ID.
* Initially, `vocab` has 1 item (`<unk>`). So, the first new word gets ID `1`.
* Now the dictionary has 2 items. The next word gets ID `2`.
* This automatically creates sequential numbers (1, 2, 3, 4...) for every new word.



```python
# Print the final dictionary.
vocab

```

* **What it does:** Outputs the completed vocabulary dictionary, which will look something like this: `{'<unk>': 0, 'the': 1, 'last': 2, 'train': 3, ...}`.

---

---

### 📖 Vocabulary কেন তৈরি করা হয়? (বাংলা ভার্সন)

ডিপ লার্নিং মডেল (যেমন LSTM) মূলত একটি বিশাল ক্যালকুলেটর। এরা সরাসরি কোনো টেক্সট বা শব্দ (যেমন "train" বা "journey") পড়তে বা বুঝতে পারে না; এরা শুধু নাম্বার (ম্যাট্রিক্স এবং টেন্সর) বোঝে।

এই কারণে, টেক্সটকে নাম্বারে কনভার্ট করার জন্য আমাদের একটি **"ডিকশনারি" বা "লুকআপ টেবিল (Lookup Table)"** তৈরি করতে হয়, যেখানে আমাদের ডেটাসেটের প্রতিটি ইউনিক শব্দের বিপরীতে একটি নির্দিষ্ট নাম্বার (ID) দেওয়া থাকে। এই ম্যাপিং বা ডিকশনারিটিকেই Vocabulary বলা হয়।

**`<unk>` টোকেন কেন লাগে?**
মডেলটি যখন বাস্তবে (Real-world) কাজ করবে, তখন ইউজার এমন কোনো শব্দ ইনপুট দিতে পারে যা মডেলটি আগে কখনো দেখেনি (যেমন "helicopter")। সেই শব্দের যদি কোনো নাম্বার বা ID না থাকে, তাহলে কোড ক্র্যাশ করবে। এই সমস্যা সমাধানের জন্যই `<unk>` (Unknown) টোকেন ব্যবহার করা হয়। মডেল যখনই কোনো অপরিচিত শব্দ পায়, তখন সে সেটিকে `<unk>`-এর ID (অর্থাৎ 0) দিয়ে রিপ্লেস করে দেয়। ফলে মডেল ক্র্যাশ না করে কাজ চালিয়ে যেতে পারে।

### 📝 কোডের লাইন-বাই-লাইন এক্সপ্ল্যানেশন

```python
# অজানা শব্দের জন্য একটি টোকেন দিয়ে ডিকশনারি শুরু করা হলো।
vocab = { '<unk>' : 0 }

```

* **কাজ:** এটি `vocab` নামের একটি পাইথন ডিকশনারি তৈরি করে। শুরুতেই এতে `'<unk>'` টোকেনটি রাখা হয় এবং এর ভ্যালু বা ID হিসেবে `0` সেট করে দেওয়া হয়।

```python
# আমাদের টেক্সটের ইউনিক শব্দগুলোর ওপর লুপ চালানো হচ্ছে।
for token in Counter(tokens).keys():

```

* **কাজ:** `Counter(tokens)` আমাদের ডেটাসেটে কোন শব্দ কতবার আছে তার হিসাব বের করে। এর সাথে `.keys()` যুক্ত করায়, ফ্রিকোয়েন্সি বা সংখ্যাগুলো বাদ গিয়ে শুধুমাত্র **ইউনিক শব্দগুলোর** একটি লিস্ট তৈরি হয়। `for` লুপটি সেই ইউনিক শব্দগুলোর লিস্ট ধরে এক এক করে সামনের দিকে এগোয়।

```python
# শব্দটি আগে থেকেই ডিকশনারিতে আছে কি না তা চেক করা।
  if token not in vocab:

```

* **কাজ:** এটি একটি সেফটি চেক। এর মাধ্যমে নিশ্চিত করা হয় যে আমরা ভুল করে আগের কোনো টোকেনকে (যেমন `<unk>`) ওভাররাইট করছি না।

```python
# শব্দটির জন্য একটি নতুন ও ইউনিক নাম্বার (ID) সেট করা।
    vocab[token] = len(vocab)

```

* **কাজ:** এটি এই কোডের সবচেয়ে বুদ্ধিদীপ্ত লাইন! এটি ডিকশনারির *বর্তমান সাইজ বা লেন্থকে* নতুন শব্দের ID হিসেবে সেট করে দেয়।
* শুরুতে ডিকশনারিতে আইটেম থাকে ১টি (`<unk>`)। তাই প্রথম নতুন শব্দের ID হয় `১`।
* এরপর আইটেম হয়ে যায় ২টি। তখন পরের শব্দের ID হয় `২`।
* এভাবে এটি স্বয়ংক্রিয়ভাবে প্রতিটি নতুন শব্দের জন্য সিরিয়াল অনুযায়ী নাম্বার (১, ২, ৩, ৪...) বসাতে থাকে।



```python
# ফাইনাল ডিকশনারিটি প্রিন্ট করা।
vocab

```

* **কাজ:** এটি আমাদের সম্পূর্ণ তৈরি হওয়া ডিকশনারিটি আউটপুট হিসেবে দেখাবে, যা দেখতে এমন হবে: `{'<unk>': 0, 'the': 1, 'last': 2, 'train': 3, ...}`।

### **build vocab**

In [7]:
vocab = { '<unk>' : 0 }

In [8]:
for token in Counter(tokens).keys():
  if token not in vocab:
    vocab[token] = len(vocab)

vocab


{'<unk>': 0,
 'the': 1,
 'last': 2,
 'train': 3,
 'journey': 4,
 'on': 5,
 'a': 6,
 'cold': 7,
 'winter': 8,
 'evening': 9,
 ',': 10,
 'daniel': 11,
 'arrived': 12,
 'at': 13,
 'old': 14,
 'railway': 15,
 'station': 16,
 'carrying': 17,
 'small': 18,
 'leather': 19,
 'bag': 20,
 '.': 21,
 'was': 22,
 'nearly': 23,
 'empty': 24,
 'except': 25,
 'for': 26,
 'few': 27,
 'passengers': 28,
 'waiting': 29,
 'quietly': 30,
 'wooden': 31,
 'benches': 32,
 'loud': 33,
 'whistle': 34,
 'echoed': 35,
 'through': 36,
 'air': 37,
 'as': 38,
 'final': 39,
 'of': 40,
 'night': 41,
 'slowly': 42,
 'entered': 43,
 'platform': 44,
 'checked': 45,
 'ticket': 46,
 'in': 47,
 'his': 48,
 'pocket': 49,
 'and': 50,
 'walked': 51,
 'toward': 52,
 'metal': 53,
 'doors': 54,
 'opened': 55,
 'with': 56,
 'creaking': 57,
 'sound': 58,
 'warm': 59,
 'rushed': 60,
 'outside': 61,
 'inside': 62,
 'yellow': 63,
 'lights': 64,
 'flickered': 65,
 'softly': 66,
 'above': 67,
 'rows': 68,
 'seats': 69,
 'he': 70,
 'chose

In [9]:
len(vocab)

284

# **Extracting Sentences and Tokenization**

### Extracting Sentences and Tokenization (English Version)

Our main task begins here. First, we need to extract individual sentences from our raw document, clean them, and then tokenize them into numerical formats. Without this, our model cannot process the text.

#### 1. Extracting and Cleaning Sentences

If we print our raw `document`, we will see newline characters (`\n`). If we just use `document.split('\n')`, we get a list of sentences, but it also includes empty strings (`''`). We need to clean this up.

```python
# Splitting by newline and filtering out empty lines in one go
input_sentences = [line for line in document.split('\n') if line.strip() != '']

# Print to verify
input_sentences

```

**Line-by-Line Explanation:**

* `document.split('\n')`: Breaks the entire story into a list of separate lines wherever there is a line break.
* `for line in ...`: Loops through that newly generated list of lines.
* `if line.strip() != ''`: Checks if a line is empty. If it has actual text, it keeps it; if it is just an empty space, it throws it away.
* `input_sentences`: Now holds our perfectly clean list of sentences (e.g., `['The Last Train Journey', 'On a cold winter evening...', ...]`).

#### 2. Tokenizing and Converting to Numerical Indices

Now that we have clean sentences, we need to convert every word in these sentences into its corresponding ID from our `vocab` dictionary.

```python
# Function to convert a single sentence into a list of numbers
def text_to_indices(sentence, vocab):
    numerical_sentence = []
    for token in sentence:
        if token in vocab:
            numerical_sentence.append(vocab[token])
        else:
            numerical_sentence.append(vocab['<unk>'])
    return numerical_sentence

# List to store all numerical sentences
input_numerical_sentecnes = []

# Loop through our cleaned sentences
for sentence in input_sentences:
    input_numerical_sentecnes.append(text_to_indices(word_tokenize(sentence.lower()), vocab))

```

**Line-by-Line Explanation:**

* `def text_to_indices(sentence, vocab):`: Defines a custom function that takes a tokenized sentence and our vocabulary dictionary as inputs.
* `numerical_sentence = []`: Creates an empty list to store the numerical IDs for the current sentence.
* `for token in sentence:`: Loops through every word (token) in the given sentence.
* `if token in vocab:`: Checks if the word exists in our vocabulary.
* `numerical_sentence.append(vocab[token])`: If it exists, it finds the ID and adds it to our list.
* `else: numerical_sentence.append(vocab['<unk>'])`: If the word is missing, it adds the ID for the unknown token (`0`) to prevent errors.
* `input_numerical_sentecnes = []`: Creates the master list to hold all the finalized numerical sentences.
* `for sentence in input_sentences:`: Loops through our clean, text-based sentences.
* `word_tokenize(sentence.lower())`: Converts the sentence to lowercase and splits it into individual words using NLTK.
* `text_to_indices(...)`: Passes those words into our custom function to get numbers, and then appends that list of numbers to `input_numerical_sentecnes`.

---

### 📝 স্টেপ ২: ডেটা থেকে বাক্য আলাদা করা এবং টোকেনাইজেশন (বাংলা ভার্সন)

আমাদের মূল প্রজেক্টের কাজ মূলত এখান থেকেই শুরু। প্রথমে আমাদের মূল ডেটাসেট থেকে বাক্যগুলোকে আলাদা করতে হবে, ক্লিন করতে হবে এবং এরপর টোকেনাইজ করে নাম্বারে কনভার্ট করতে হবে। এটি ছাড়া মডেল কাজ করতে পারবে না।

#### ১. বাক্য আলাদা করা এবং ক্লিন করা

আমরা যদি `document` প্রিন্ট করি, তবে দেখব সেখানে `\n` আছে। আমরা যদি শুধু `document.split('\n')` ব্যবহার করি, তাহলে আমরা লিস্ট পাব ঠিকই, কিন্তু সেখানে অনেক ফাঁকা স্ট্রিং (`''`) থেকে যাবে। আমাদের এটি ক্লিন করতে হবে।

```python
# লাইন ব্রেক অনুযায়ী স্প্লিট করা এবং একই সাথে ফাঁকা লাইন মুছে ফেলা
input_sentences = [line for line in document.split('\n') if line.strip() != '']

# আউটপুট চেক করা
input_sentences

```

**কোডের লাইন-বাই-লাইন এক্সপ্ল্যানেশন:**

* `document.split('\n')`: এটি পুরো গল্পটিকে লাইন ব্রেক বা এন্টার অনুযায়ী ভেঙে একটি লিস্ট তৈরি করে।
* `for line in ...`: এটি সেই নতুন তৈরি হওয়া লিস্টের প্রতিটি লাইনের ওপর লুপ চালায়।
* `if line.strip() != ''`: এটি চেক করে লাইনটিতে কোনো টেক্সট আছে কি না। ফাঁকা লাইনগুলোকে এটি ফিল্টার করে বাদ দিয়ে দেয়।
* `input_sentences`: এর ভেতরে এখন আমাদের একদম ফ্রেশ এবং ক্লিন বাক্যগুলোর লিস্ট জমা হয়ে গেছে।

#### ২. টোকেনাইজ করা এবং নাম্বারে রূপান্তর (Indices)

যেহেতু আমাদের কাছে ক্লিন বাক্য আছে, এখন আমাদের কাজ হলো এই বাক্যগুলোর প্রতিটি শব্দকে `vocab` ডিকশনারি ব্যবহার করে নাম্বারে রূপান্তর করা।

```python
# একটি বাক্যকে নাম্বারের লিস্টে কনভার্ট করার ফাংশন
def text_to_indices(sentence, vocab):
    numerical_sentence = []
    for token in sentence:
        if token in vocab:
            numerical_sentence.append(vocab[token])
        else:
            numerical_sentence.append(vocab['<unk>'])
    return numerical_sentence

# সব বাক্যের নাম্বার সেভ করার জন্য মেইন লিস্ট
input_numerical_sentecnes = []

# আমাদের ক্লিন করা বাক্যগুলোর ওপর লুপ চালানো
for sentence in input_sentences:
    input_numerical_sentecnes.append(text_to_indices(word_tokenize(sentence.lower()), vocab))

```

**কোডের লাইন-বাই-লাইন এক্সপ্ল্যানেশন:**

* `def text_to_indices(sentence, vocab):`: এটি একটি কাস্টম ফাংশন তৈরি করে, যার কাজ হলো শব্দ গ্রহণ করে নাম্বার ফেরত দেওয়া।
* `numerical_sentence = []`: বর্তমান বাক্যের নাম্বারগুলো রাখার জন্য একটি ফাঁকা লিস্ট তৈরি করে।
* `for token in sentence:`: বাক্যের প্রতিটি শব্দের ওপর লুপ চালায়।
* `if token in vocab:`: চেক করে শব্দটি আমাদের `vocab` ডিকশনারিতে আছে কি না।
* `numerical_sentence.append(vocab[token])`: যদি থাকে, তবে ডিকশনারি থেকে তার নাম্বার (ID) নিয়ে লিস্টে যুক্ত করে।
* `else: numerical_sentence.append(vocab['<unk>'])`: যদি না থাকে, তবে মডেল যেন ক্র্যাশ না করে তাই অজানা শব্দের ID (`0`) যুক্ত করে দেয়।
* `input_numerical_sentecnes = []`: সবগুলো বাক্যের ফাইনাল রেজাল্ট রাখার জন্য একটি মেইন লিস্ট তৈরি করে।
* `for sentence in input_sentences:`: আমাদের ক্লিন করা বাক্যগুলোর ওপর লুপ চালায়।
* `word_tokenize(sentence.lower())`: বাক্যটিকে ছোট হাতের অক্ষরে (lowercase) কনভার্ট করে এবং NLTK দিয়ে আলাদা আলাদা শব্দে ভাঙে।
* `text_to_indices(...)`: সেই শব্দগুলোকে আমাদের তৈরি করা ফাংশনে পাঠায় এবং প্রাপ্ত নাম্বারের লিস্টটিকে মেইন লিস্টে (`input_numerical_sentecnes`) যুক্ত করে দেয়।

### **Raw Code Practice Purpose**

In [10]:
document

'The Last Train Journey\n\nOn a cold winter evening, Daniel arrived at the old railway station carrying a small leather bag.\nThe station was nearly empty except for a few passengers waiting quietly on wooden benches.\nA loud whistle echoed through the air as the final train of the night slowly entered the platform.\n\nDaniel checked the ticket in his pocket and walked toward the train.\nThe metal doors opened with a creaking sound and warm air rushed outside.\nInside the train, yellow lights flickered softly above rows of empty seats.\n\nHe chose a seat beside the window and placed his bag carefully near his feet.\nAs the train began to move, the city lights slowly disappeared behind thick clouds of fog.\nSnow started falling gently outside while distant mountains appeared under the moonlight.\n\nAcross the aisle sat an elderly woman reading an old book with a blue cover.\nEvery few minutes she looked outside the window as if searching for something familiar.\nDaniel noticed a silver 

In [11]:
document.split('\n')

['The Last Train Journey',
 '',
 'On a cold winter evening, Daniel arrived at the old railway station carrying a small leather bag.',
 'The station was nearly empty except for a few passengers waiting quietly on wooden benches.',
 'A loud whistle echoed through the air as the final train of the night slowly entered the platform.',
 '',
 'Daniel checked the ticket in his pocket and walked toward the train.',
 'The metal doors opened with a creaking sound and warm air rushed outside.',
 'Inside the train, yellow lights flickered softly above rows of empty seats.',
 '',
 'He chose a seat beside the window and placed his bag carefully near his feet.',
 'As the train began to move, the city lights slowly disappeared behind thick clouds of fog.',
 'Snow started falling gently outside while distant mountains appeared under the moonlight.',
 '',
 'Across the aisle sat an elderly woman reading an old book with a blue cover.',
 'Every few minutes she looked outside the window as if searching for

In [12]:
input_sentences = document.split('\n')

In [13]:
input_sentences[1]

''

In [14]:
# small preprocessing
# Splitting by newline and filtering out empty lines in one go
input_sentences = [line for line in document.split('\n') if line.strip() != '']

# Print to verify
input_sentences

['The Last Train Journey',
 'On a cold winter evening, Daniel arrived at the old railway station carrying a small leather bag.',
 'The station was nearly empty except for a few passengers waiting quietly on wooden benches.',
 'A loud whistle echoed through the air as the final train of the night slowly entered the platform.',
 'Daniel checked the ticket in his pocket and walked toward the train.',
 'The metal doors opened with a creaking sound and warm air rushed outside.',
 'Inside the train, yellow lights flickered softly above rows of empty seats.',
 'He chose a seat beside the window and placed his bag carefully near his feet.',
 'As the train began to move, the city lights slowly disappeared behind thick clouds of fog.',
 'Snow started falling gently outside while distant mountains appeared under the moonlight.',
 'Across the aisle sat an elderly woman reading an old book with a blue cover.',
 'Every few minutes she looked outside the window as if searching for something familiar.

In [15]:
input_sentences[1]

'On a cold winter evening, Daniel arrived at the old railway station carrying a small leather bag.'

In [16]:
for sentence in input_sentences:
  print(word_tokenize(sentence.lower()))

['the', 'last', 'train', 'journey']
['on', 'a', 'cold', 'winter', 'evening', ',', 'daniel', 'arrived', 'at', 'the', 'old', 'railway', 'station', 'carrying', 'a', 'small', 'leather', 'bag', '.']
['the', 'station', 'was', 'nearly', 'empty', 'except', 'for', 'a', 'few', 'passengers', 'waiting', 'quietly', 'on', 'wooden', 'benches', '.']
['a', 'loud', 'whistle', 'echoed', 'through', 'the', 'air', 'as', 'the', 'final', 'train', 'of', 'the', 'night', 'slowly', 'entered', 'the', 'platform', '.']
['daniel', 'checked', 'the', 'ticket', 'in', 'his', 'pocket', 'and', 'walked', 'toward', 'the', 'train', '.']
['the', 'metal', 'doors', 'opened', 'with', 'a', 'creaking', 'sound', 'and', 'warm', 'air', 'rushed', 'outside', '.']
['inside', 'the', 'train', ',', 'yellow', 'lights', 'flickered', 'softly', 'above', 'rows', 'of', 'empty', 'seats', '.']
['he', 'chose', 'a', 'seat', 'beside', 'the', 'window', 'and', 'placed', 'his', 'bag', 'carefully', 'near', 'his', 'feet', '.']
['as', 'the', 'train', 'began

In [17]:
input_numerical_sentecnes = []
for sentence in input_sentences:
  input_numerical_sentecnes.append(word_tokenize(sentence.lower()))

In [18]:
input_numerical_sentecnes

[['the', 'last', 'train', 'journey'],
 ['on',
  'a',
  'cold',
  'winter',
  'evening',
  ',',
  'daniel',
  'arrived',
  'at',
  'the',
  'old',
  'railway',
  'station',
  'carrying',
  'a',
  'small',
  'leather',
  'bag',
  '.'],
 ['the',
  'station',
  'was',
  'nearly',
  'empty',
  'except',
  'for',
  'a',
  'few',
  'passengers',
  'waiting',
  'quietly',
  'on',
  'wooden',
  'benches',
  '.'],
 ['a',
  'loud',
  'whistle',
  'echoed',
  'through',
  'the',
  'air',
  'as',
  'the',
  'final',
  'train',
  'of',
  'the',
  'night',
  'slowly',
  'entered',
  'the',
  'platform',
  '.'],
 ['daniel',
  'checked',
  'the',
  'ticket',
  'in',
  'his',
  'pocket',
  'and',
  'walked',
  'toward',
  'the',
  'train',
  '.'],
 ['the',
  'metal',
  'doors',
  'opened',
  'with',
  'a',
  'creaking',
  'sound',
  'and',
  'warm',
  'air',
  'rushed',
  'outside',
  '.'],
 ['inside',
  'the',
  'train',
  ',',
  'yellow',
  'lights',
  'flickered',
  'softly',
  'above',
  'rows',
  '

In [19]:
def text_to_indices(sentence, vocab):
  numerical_sentence = []
  for token in sentence:
    if token in vocab:
      numerical_sentence.append(vocab[token])
    else:
      numerical_sentence.append(vocab['<unk>'])
  return numerical_sentence


In [20]:
input_numerical_sentecnes = []

for sentence in input_sentences:
  input_numerical_sentecnes.append(text_to_indices(word_tokenize(sentence.lower()), vocab))

input_numerical_sentecnes

[[1, 2, 3, 4],
 [5, 6, 7, 8, 9, 10, 11, 12, 13, 1, 14, 15, 16, 17, 6, 18, 19, 20, 21],
 [1, 16, 22, 23, 24, 25, 26, 6, 27, 28, 29, 30, 5, 31, 32, 21],
 [6, 33, 34, 35, 36, 1, 37, 38, 1, 39, 3, 40, 1, 41, 42, 43, 1, 44, 21],
 [11, 45, 1, 46, 47, 48, 49, 50, 51, 52, 1, 3, 21],
 [1, 53, 54, 55, 56, 6, 57, 58, 50, 59, 37, 60, 61, 21],
 [62, 1, 3, 10, 63, 64, 65, 66, 67, 68, 40, 24, 69, 21],
 [70, 71, 6, 72, 73, 1, 74, 50, 75, 48, 20, 76, 77, 48, 78, 21],
 [38, 1, 3, 79, 80, 81, 10, 1, 82, 64, 42, 83, 84, 85, 86, 40, 87, 21],
 [88, 89, 90, 91, 61, 92, 93, 94, 95, 96, 1, 97, 21],
 [98, 1, 99, 100, 101, 102, 103, 104, 101, 14, 105, 56, 6, 106, 107, 21],
 [108, 27, 109, 110, 111, 61, 1, 74, 38, 112, 113, 26, 114, 115, 21],
 [11, 116, 6, 117, 118, 119, 120, 121, 122, 123, 6, 18, 124, 21],
 [125, 126, 127, 10, 1, 3, 128, 13, 6, 129, 16, 130, 131, 132, 133, 134, 21],
 [135, 27, 136, 43, 137, 138, 1, 3, 13, 139, 140, 21],
 [6, 141, 142, 143, 1, 16, 38, 1, 54, 144, 145, 146, 21],
 [147, 148, 1, 103

### **Extract sentence from data**

In [21]:
input_sentences = document.split('\n')
input_sentences

['The Last Train Journey',
 '',
 'On a cold winter evening, Daniel arrived at the old railway station carrying a small leather bag.',
 'The station was nearly empty except for a few passengers waiting quietly on wooden benches.',
 'A loud whistle echoed through the air as the final train of the night slowly entered the platform.',
 '',
 'Daniel checked the ticket in his pocket and walked toward the train.',
 'The metal doors opened with a creaking sound and warm air rushed outside.',
 'Inside the train, yellow lights flickered softly above rows of empty seats.',
 '',
 'He chose a seat beside the window and placed his bag carefully near his feet.',
 'As the train began to move, the city lights slowly disappeared behind thick clouds of fog.',
 'Snow started falling gently outside while distant mountains appeared under the moonlight.',
 '',
 'Across the aisle sat an elderly woman reading an old book with a blue cover.',
 'Every few minutes she looked outside the window as if searching for

**small preprocessing**

In [22]:
# Splitting by newline and filtering out empty lines in one go
input_sentences = [line for line in document.split('\n') if line.strip() != '']

# Print to verify
input_sentences

['The Last Train Journey',
 'On a cold winter evening, Daniel arrived at the old railway station carrying a small leather bag.',
 'The station was nearly empty except for a few passengers waiting quietly on wooden benches.',
 'A loud whistle echoed through the air as the final train of the night slowly entered the platform.',
 'Daniel checked the ticket in his pocket and walked toward the train.',
 'The metal doors opened with a creaking sound and warm air rushed outside.',
 'Inside the train, yellow lights flickered softly above rows of empty seats.',
 'He chose a seat beside the window and placed his bag carefully near his feet.',
 'As the train began to move, the city lights slowly disappeared behind thick clouds of fog.',
 'Snow started falling gently outside while distant mountains appeared under the moonlight.',
 'Across the aisle sat an elderly woman reading an old book with a blue cover.',
 'Every few minutes she looked outside the window as if searching for something familiar.

In [23]:
# Function to convert a single sentence into a list of numbers
def text_to_indices(sentence, vocab):
  numerical_sentence = []
  for token in sentence:
    if token in vocab:
      numerical_sentence.append(vocab[token])
    else:
      numerical_sentence.append(vocab['<unk>'])
  return numerical_sentence

In [24]:
# List to store all numerical sentences
input_numerical_sentecnes = []

# Loop through our cleaned sentences
for sentence in input_sentences:
  input_numerical_sentecnes.append(text_to_indices(word_tokenize(sentence.lower()), vocab))

input_numerical_sentecnes

[[1, 2, 3, 4],
 [5, 6, 7, 8, 9, 10, 11, 12, 13, 1, 14, 15, 16, 17, 6, 18, 19, 20, 21],
 [1, 16, 22, 23, 24, 25, 26, 6, 27, 28, 29, 30, 5, 31, 32, 21],
 [6, 33, 34, 35, 36, 1, 37, 38, 1, 39, 3, 40, 1, 41, 42, 43, 1, 44, 21],
 [11, 45, 1, 46, 47, 48, 49, 50, 51, 52, 1, 3, 21],
 [1, 53, 54, 55, 56, 6, 57, 58, 50, 59, 37, 60, 61, 21],
 [62, 1, 3, 10, 63, 64, 65, 66, 67, 68, 40, 24, 69, 21],
 [70, 71, 6, 72, 73, 1, 74, 50, 75, 48, 20, 76, 77, 48, 78, 21],
 [38, 1, 3, 79, 80, 81, 10, 1, 82, 64, 42, 83, 84, 85, 86, 40, 87, 21],
 [88, 89, 90, 91, 61, 92, 93, 94, 95, 96, 1, 97, 21],
 [98, 1, 99, 100, 101, 102, 103, 104, 101, 14, 105, 56, 6, 106, 107, 21],
 [108, 27, 109, 110, 111, 61, 1, 74, 38, 112, 113, 26, 114, 115, 21],
 [11, 116, 6, 117, 118, 119, 120, 121, 122, 123, 6, 18, 124, 21],
 [125, 126, 127, 10, 1, 3, 128, 13, 6, 129, 16, 130, 131, 132, 133, 134, 21],
 [135, 27, 136, 43, 137, 138, 1, 3, 13, 139, 140, 21],
 [6, 141, 142, 143, 1, 16, 38, 1, 54, 144, 145, 146, 21],
 [147, 148, 1, 103

# **📝 Forming Training Sequences**

Now we need to form our training sequences. For a supervised learning model to predict the next word, it needs to see how a sentence builds up word by word.

If our numerical sentence is `[1, 2, 3, 4]`, we cannot just feed the whole array to the model at once. We must break it down into smaller, growing sequences like this:

* `[1, 2]` (Model sees 1, predicts 2)
* `[1, 2, 3]` (Model sees 1 and 2, predicts 3)
* `[1, 2, 3, 4]` (Model sees 1, 2, and 3, predicts 4)

**The Final Code:**

```python
# Create an empty list to store all our generated sequences
training_sequence = []

# Loop through each numerical sentence
for sentence in input_numerical_sentecnes:
    # Loop through the indices of the sentence, starting from index 1
    for i in range(1, len(sentence)):
        # Slice the sentence from the start up to the current index + 1
        training_sequence.append(sentence[:i+1])

# Check the generated sequences
training_sequence

```

**Line-by-Line Explanation:**

* `training_sequence = []`: Creates an empty master list where all our newly sliced sequences will be stored.
* `for sentence in input_numerical_sentecnes:`: This outer loop goes through every single full numerical sentence in our dataset.
* `for i in range(1, len(sentence)):`: This inner loop controls the slicing. Notice that it starts at `1`, not `0`. Why? Because a sequence needs at least two words (one as input/feature, one as target). If we started at `0`, our slice would just be `[1]`, which has no target word to predict.
* `sentence[:i+1]`: This is Python list slicing. It grabs all numbers from the beginning of the list (`0`) up to the position `i`. As `i` increases, the slice gets longer.
* `training_sequence.append(...)`: Adds that freshly sliced piece of the sentence into our master list.

---

### 📝 স্টেপ ৩: ট্রেনিং সিকোয়েন্স তৈরি করা (বাংলা ভার্সন)

এবার আমাদের ট্রেনিং সিকোয়েন্স তৈরি করার পালা। একটি মডেলকে দিয়ে পরবর্তী শব্দ প্রেডিক্ট করাতে হলে, তাকে শেখাতে হয় যে একটি বাক্য কীভাবে শব্দে শব্দে বড় হয়।

আমাদের নাম্বারের লিস্ট যদি হয় `[1, 2, 3, 4]`, তবে আমরা পুরো লিস্টটা একবারে মডেলকে দিতে পারব না। আমাদের এটিকে ভেঙে ছোট ছোট ক্রমবর্ধমান সিকোয়েন্স তৈরি করতে হবে:

* `[1, 2]` (মডেল 1 দেখবে, 2 প্রেডিক্ট করবে)
* `[1, 2, 3]` (মডেল 1, 2 দেখবে, 3 প্রেডিক্ট করবে)
* `[1, 2, 3, 4]` (মডেল 1, 2, 3 দেখবে, 4 প্রেডিক্ট করবে)

**ফাইনাল কোড:**

```python
# সব সিকোয়েন্স সেভ করার জন্য একটি ফাঁকা লিস্ট তৈরি
training_sequence = []

# প্রতিটি নাম্বারের লিস্ট বা বাক্যের ওপর লুপ চালানো
for sentence in input_numerical_sentecnes:
    # ইনডেক্স ১ থেকে বাক্যের শেষ পর্যন্ত লুপ চালানো
    for i in range(1, len(sentence)):
        # বাক্যের শুরু থেকে বর্তমান ইনডেক্স+১ পর্যন্ত কেটে নেওয়া
        training_sequence.append(sentence[:i+1])

# জেনারেট হওয়া সিকোয়েন্সগুলো চেক করা
training_sequence

```

**কোডের লাইন-বাই-লাইন এক্সপ্ল্যানেশন:**

* `training_sequence = []`: এটি আমাদের মূল ফাঁকা লিস্ট, যেখানে কেটে নেওয়া ছোট ছোট সিকোয়েন্সগুলো জমা হবে।
* `for sentence in input_numerical_sentecnes:`: এই আউটার লুপটি আমাদের ডেটাসেটের প্রতিটি সম্পূর্ণ নাম্বারের বাক্যের ওপর চলবে।
* `for i in range(1, len(sentence)):`: এই ইনার লুপটি মূলত বাক্য কাটার (Slicing) কাজ কন্ট্রোল করে। খেয়াল করলে দেখবে, এটি `0` থেকে শুরু না হয়ে `1` থেকে শুরু হয়েছে। কারণ কী? কারণ একটি সিকোয়েন্সে কমপক্ষে দুটি শব্দ থাকতে হয় (একটি ইনপুট, অন্যটি টার্গেট)। যদি লুপ `0` থেকে শুরু হতো, তবে আমরা শুধু `[1]` পেতাম, যেখানে প্রেডিক্ট করার মতো কোনো টার্গেট নেই।
* `sentence[:i+1]`: এটি পাইথনের লিস্ট স্লাইসিং (Slicing)। এটি লিস্টের একদম শুরু থেকে `i` পজিশন পর্যন্ত নাম্বারগুলো কেটে নেয়। `i`-এর মান যত বাড়ে, কেটে নেওয়া অংশটিও তত বড় হতে থাকে।
* `training_sequence.append(...)`: কেটে নেওয়া সিকোয়েন্সটিকে আমাদের মূল লিস্টে যুক্ত করে দেয়।

### **Raw Code Practice Purpose**

In [25]:
for sentence in input_numerical_sentecnes:
  print(sentence)

[1, 2, 3, 4]
[5, 6, 7, 8, 9, 10, 11, 12, 13, 1, 14, 15, 16, 17, 6, 18, 19, 20, 21]
[1, 16, 22, 23, 24, 25, 26, 6, 27, 28, 29, 30, 5, 31, 32, 21]
[6, 33, 34, 35, 36, 1, 37, 38, 1, 39, 3, 40, 1, 41, 42, 43, 1, 44, 21]
[11, 45, 1, 46, 47, 48, 49, 50, 51, 52, 1, 3, 21]
[1, 53, 54, 55, 56, 6, 57, 58, 50, 59, 37, 60, 61, 21]
[62, 1, 3, 10, 63, 64, 65, 66, 67, 68, 40, 24, 69, 21]
[70, 71, 6, 72, 73, 1, 74, 50, 75, 48, 20, 76, 77, 48, 78, 21]
[38, 1, 3, 79, 80, 81, 10, 1, 82, 64, 42, 83, 84, 85, 86, 40, 87, 21]
[88, 89, 90, 91, 61, 92, 93, 94, 95, 96, 1, 97, 21]
[98, 1, 99, 100, 101, 102, 103, 104, 101, 14, 105, 56, 6, 106, 107, 21]
[108, 27, 109, 110, 111, 61, 1, 74, 38, 112, 113, 26, 114, 115, 21]
[11, 116, 6, 117, 118, 119, 120, 121, 122, 123, 6, 18, 124, 21]
[125, 126, 127, 10, 1, 3, 128, 13, 6, 129, 16, 130, 131, 132, 133, 134, 21]
[135, 27, 136, 43, 137, 138, 1, 3, 13, 139, 140, 21]
[6, 141, 142, 143, 1, 16, 38, 1, 54, 144, 145, 146, 21]
[147, 148, 1, 103, 10, 11, 149, 89, 6, 150, 21]
[1

In [26]:
for sentence in input_numerical_sentecnes:
  for i in range(1, len(sentence)):
    print(sentence[:i+1])

[1, 2]
[1, 2, 3]
[1, 2, 3, 4]
[5, 6]
[5, 6, 7]
[5, 6, 7, 8]
[5, 6, 7, 8, 9]
[5, 6, 7, 8, 9, 10]
[5, 6, 7, 8, 9, 10, 11]
[5, 6, 7, 8, 9, 10, 11, 12]
[5, 6, 7, 8, 9, 10, 11, 12, 13]
[5, 6, 7, 8, 9, 10, 11, 12, 13, 1]
[5, 6, 7, 8, 9, 10, 11, 12, 13, 1, 14]
[5, 6, 7, 8, 9, 10, 11, 12, 13, 1, 14, 15]
[5, 6, 7, 8, 9, 10, 11, 12, 13, 1, 14, 15, 16]
[5, 6, 7, 8, 9, 10, 11, 12, 13, 1, 14, 15, 16, 17]
[5, 6, 7, 8, 9, 10, 11, 12, 13, 1, 14, 15, 16, 17, 6]
[5, 6, 7, 8, 9, 10, 11, 12, 13, 1, 14, 15, 16, 17, 6, 18]
[5, 6, 7, 8, 9, 10, 11, 12, 13, 1, 14, 15, 16, 17, 6, 18, 19]
[5, 6, 7, 8, 9, 10, 11, 12, 13, 1, 14, 15, 16, 17, 6, 18, 19, 20]
[5, 6, 7, 8, 9, 10, 11, 12, 13, 1, 14, 15, 16, 17, 6, 18, 19, 20, 21]
[1, 16]
[1, 16, 22]
[1, 16, 22, 23]
[1, 16, 22, 23, 24]
[1, 16, 22, 23, 24, 25]
[1, 16, 22, 23, 24, 25, 26]
[1, 16, 22, 23, 24, 25, 26, 6]
[1, 16, 22, 23, 24, 25, 26, 6, 27]
[1, 16, 22, 23, 24, 25, 26, 6, 27, 28]
[1, 16, 22, 23, 24, 25, 26, 6, 27, 28, 29]
[1, 16, 22, 23, 24, 25, 26, 6, 27, 28, 

In [27]:
# Training Sequece
training_sequence = []
for sentence in input_numerical_sentecnes:
  for i in range(1, len(sentence)):
    training_sequence.append(sentence[:i+1])

In [28]:
training_sequence

[[1, 2],
 [1, 2, 3],
 [1, 2, 3, 4],
 [5, 6],
 [5, 6, 7],
 [5, 6, 7, 8],
 [5, 6, 7, 8, 9],
 [5, 6, 7, 8, 9, 10],
 [5, 6, 7, 8, 9, 10, 11],
 [5, 6, 7, 8, 9, 10, 11, 12],
 [5, 6, 7, 8, 9, 10, 11, 12, 13],
 [5, 6, 7, 8, 9, 10, 11, 12, 13, 1],
 [5, 6, 7, 8, 9, 10, 11, 12, 13, 1, 14],
 [5, 6, 7, 8, 9, 10, 11, 12, 13, 1, 14, 15],
 [5, 6, 7, 8, 9, 10, 11, 12, 13, 1, 14, 15, 16],
 [5, 6, 7, 8, 9, 10, 11, 12, 13, 1, 14, 15, 16, 17],
 [5, 6, 7, 8, 9, 10, 11, 12, 13, 1, 14, 15, 16, 17, 6],
 [5, 6, 7, 8, 9, 10, 11, 12, 13, 1, 14, 15, 16, 17, 6, 18],
 [5, 6, 7, 8, 9, 10, 11, 12, 13, 1, 14, 15, 16, 17, 6, 18, 19],
 [5, 6, 7, 8, 9, 10, 11, 12, 13, 1, 14, 15, 16, 17, 6, 18, 19, 20],
 [5, 6, 7, 8, 9, 10, 11, 12, 13, 1, 14, 15, 16, 17, 6, 18, 19, 20, 21],
 [1, 16],
 [1, 16, 22],
 [1, 16, 22, 23],
 [1, 16, 22, 23, 24],
 [1, 16, 22, 23, 24, 25],
 [1, 16, 22, 23, 24, 25, 26],
 [1, 16, 22, 23, 24, 25, 26, 6],
 [1, 16, 22, 23, 24, 25, 26, 6, 27],
 [1, 16, 22, 23, 24, 25, 26, 6, 27, 28],
 [1, 16, 22, 23, 24, 2

**Form training sequence**

In [29]:
# Training Sequece
training_sequence = []
for sentence in input_numerical_sentecnes:
  for i in range(1, len(sentence)):
    training_sequence.append(sentence[:i+1])

In [30]:
training_sequence

[[1, 2],
 [1, 2, 3],
 [1, 2, 3, 4],
 [5, 6],
 [5, 6, 7],
 [5, 6, 7, 8],
 [5, 6, 7, 8, 9],
 [5, 6, 7, 8, 9, 10],
 [5, 6, 7, 8, 9, 10, 11],
 [5, 6, 7, 8, 9, 10, 11, 12],
 [5, 6, 7, 8, 9, 10, 11, 12, 13],
 [5, 6, 7, 8, 9, 10, 11, 12, 13, 1],
 [5, 6, 7, 8, 9, 10, 11, 12, 13, 1, 14],
 [5, 6, 7, 8, 9, 10, 11, 12, 13, 1, 14, 15],
 [5, 6, 7, 8, 9, 10, 11, 12, 13, 1, 14, 15, 16],
 [5, 6, 7, 8, 9, 10, 11, 12, 13, 1, 14, 15, 16, 17],
 [5, 6, 7, 8, 9, 10, 11, 12, 13, 1, 14, 15, 16, 17, 6],
 [5, 6, 7, 8, 9, 10, 11, 12, 13, 1, 14, 15, 16, 17, 6, 18],
 [5, 6, 7, 8, 9, 10, 11, 12, 13, 1, 14, 15, 16, 17, 6, 18, 19],
 [5, 6, 7, 8, 9, 10, 11, 12, 13, 1, 14, 15, 16, 17, 6, 18, 19, 20],
 [5, 6, 7, 8, 9, 10, 11, 12, 13, 1, 14, 15, 16, 17, 6, 18, 19, 20, 21],
 [1, 16],
 [1, 16, 22],
 [1, 16, 22, 23],
 [1, 16, 22, 23, 24],
 [1, 16, 22, 23, 24, 25],
 [1, 16, 22, 23, 24, 25, 26],
 [1, 16, 22, 23, 24, 25, 26, 6],
 [1, 16, 22, 23, 24, 25, 26, 6, 27],
 [1, 16, 22, 23, 24, 25, 26, 6, 27, 28],
 [1, 16, 22, 23, 24, 2

# **Padding training sequence**

#### ⚠️ The Problem: Variable Sequence Lengths

If we look at the first 5 sequences we generated:
`[[1, 2], [1, 2, 3], [1, 2, 3, 4], [5, 6], [5, 6, 7]]`

We immediately see a major issue: **the lengths of these lists are not the same.** Neural networks (like LSTMs) process data in "batches" (e.g., 32 sequences at a time) to train efficiently. A batch is essentially a matrix, and a matrix must have perfectly equal rows and columns. If our sequences have different lengths, the model will throw an error because it cannot form a uniform batch.

#### 💡 The Solution: Pre-Padding with Zeros

To fix this, we need to make all sequences as long as the longest sequence in our dataset. We do this by adding `0`s to the beginning of the shorter sequences. This technique is called **Pre-padding**. The model ignores these `0`s (since `0` is our `<unk>` token, or acts as neutral padding), but it allows the data to fit perfectly into a matrix.

**The Code:**

```python
# Step A: Find the maximum sequence length
len_list = []
for sequence in training_sequence:
    len_list.append(len(sequence))

max_len = max(len_list)

# Step B: Apply padding to all sequences
padded_sequence = []
for sequence in training_sequence:
    # Generate the required number of zeros and add them to the front
    padded_sequence.append([0] * (max_len - len(sequence)) + sequence)

# Check the final result
padded_sequence

```

**Line-by-Line Explanation:**

* `len_list = []`: Creates an empty list to store the lengths of all our sequences.
* `len_list.append(len(sequence))`: Calculates the length of each sequence and adds it to our list.
* `max_len = max(len_list)`: Finds the largest number in that list. This is the length of our longest sentence.
* `padded_sequence = []`: Creates a master list for our final, equal-sized sequences.
* `[0] * (max_len - len(sequence))`: This is brilliant Python logic! It subtracts the current sequence's length from the `max_len` to find out how many zeros are missing. Then, it multiplies a list containing zero `[0]` by that number to generate exactly that many zeros (e.g., `[0] * 3` becomes `[0, 0, 0]`).
* `... + sequence`: Concatenates (joins) those generated zeros to the *front* of the original sequence.

---

### 🛠️ ট্রেনিং সিকোয়েন্স প্যাডিং করা (বাংলা ভার্সন)

#### ⚠️ সমস্যা: সিকোয়েন্সের ভিন্ন সাইজ বা দৈর্ঘ্য

আমরা যদি আমাদের তৈরি করা প্রথম ৫টি সিকোয়েন্সের দিকে তাকাই:
`[[1, 2], [1, 2, 3], [1, 2, 3, 4], [5, 6], [5, 6, 7]]`

আমরা খুব সহজেই একটি বড় সমস্যা দেখতে পাব: **লিস্টগুলোর সাইজ বা দৈর্ঘ্য সমান নয়।**
নিউরাল নেটওয়ার্ক (যেমন LSTM) দ্রুত কাজ করার জন্য ডেটাকে "ব্যাচ" (Batch) আকারে প্রসেস করে (যেমন: একসাথে ৩২টি সিকোয়েন্স পাঠানো)। একটি ব্যাচ মূলত একটি ম্যাট্রিক্স, যার সারি (row) এবং কলাম (column) একদম সমান হতে হয়। আমাদের সিকোয়েন্সগুলো ছোট-বড় হলে মডেল কোনো ম্যাট্রিক্স তৈরি করতে পারবে না এবং এরর (error) দেবে।

#### 💡 সমাধান: শুরুতে শূন্য (0) যোগ করা (Pre-Padding)

এই সমস্যার সমাধানের জন্য, আমাদের ডেটাসেটের সবচেয়ে বড় সিকোয়েন্সটি খুঁজে বের করতে হবে এবং বাকি ছোট সিকোয়েন্সগুলোর শুরুতে প্রয়োজনীয় সংখ্যক `0` (শূন্য) বসিয়ে সবগুলোকে সমান সাইজে নিয়ে আসতে হবে। একে বলা হয় **Pre-padding**। মডেল এই শূন্যগুলোকে ইগনোর করে, কিন্তু এর ফলে ডেটাগুলো সুন্দরভাবে ম্যাট্রিক্সে সেট হয়ে যায়।

**কোড:**

```python
# ধাপ ক: সবচেয়ে বড় সিকোয়েন্সের সাইজ বের করা
len_list = []
for sequence in training_sequence:
    len_list.append(len(sequence))

max_len = max(len_list)

# ধাপ খ: সবগুলো সিকোয়েন্সে প্যাডিং অ্যাপ্লাই করা
padded_sequence = []
for sequence in training_sequence:
    # প্রয়োজনীয় সংখ্যক শূন্য তৈরি করে সিকোয়েন্সের শুরুতে যোগ করা
    padded_sequence.append([0] * (max_len - len(sequence)) + sequence)

# ফাইনাল রেজাল্ট চেক করা
padded_sequence

```

**কোডের লাইন-বাই-লাইন এক্সপ্ল্যানেশন:**

* `len_list = []`: আমাদের সব সিকোয়েন্সের সাইজ বা দৈর্ঘ্যগুলো সেভ করার জন্য একটি লিস্ট তৈরি করে।
* `len_list.append(len(sequence))`: লুপের মাধ্যমে প্রতিটি সিকোয়েন্সের সাইজ মাপে এবং লিস্টে যুক্ত করে।
* `max_len = max(len_list)`: লিস্ট থেকে সর্বোচ্চ সাইজ বা সবচেয়ে বড় নাম্বারটি খুঁজে বের করে। এটিই হলো আমাদের টার্গেট সাইজ।
* `padded_sequence = []`: প্যাডিং করা সমান সাইজের সিকোয়েন্সগুলো রাখার জন্য একটি মেইন লিস্ট তৈরি করে।
* `[0] * (max_len - len(sequence))`: এটি চমৎকার একটি পাইথন লজিক! এটি প্রথমে `max_len` থেকে বর্তমান সিকোয়েন্সের সাইজ বিয়োগ করে দেখে যে কয়টি শূন্য শর্ট আছে। এরপর `[0]`-কে সেই সংখ্যা দিয়ে গুণ করে ঠিক ততগুলো শূন্য তৈরি করে ফেলে (যেমন: `[0] * 3` দিলে পাইথন `[0, 0, 0]` তৈরি করে)।
* `... + sequence`: তৈরি করা শূন্যগুলোর সাথে অরিজিনাল সিকোয়েন্সটিকে যুক্ত করে দেয়। যেহেতু শূন্যগুলো আগে আছে, তাই এটি সিকোয়েন্সের শুরুতেই বসে যায়।

### **Raw Code Practice Purpose**

In [31]:
len(training_sequence)

555

In [32]:
training_sequence[:5]

[[1, 2], [1, 2, 3], [1, 2, 3, 4], [5, 6], [5, 6, 7]]

In [33]:
len_list = []
for sequence in training_sequence:
  len_list.append(len(sequence))

In [34]:
max_len = max(len_list)
max_len

25

In [35]:
padded_sequence = []
for sequence in training_sequence:
  padded_sequence.append([0] * (max_len - len(sequence)) + sequence)


In [36]:
padded_sequence

[[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 2],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 2, 3],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 2, 3, 4],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 5, 6],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 5, 6, 7],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 5, 6, 7, 8],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 5, 6, 7, 8, 9],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 5, 6, 7, 8, 9, 10],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 5, 6, 7, 8, 9, 10, 11],
 [0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  5,
  6,
  7,
  8,
  9,
  10,
  11,
  12],
 [0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  5,
  6,
  7,
  8,
  9,
  10,
  11,
  12,
  13],
 [0,
  0,
  0,
  0,
  0,
  0,
  0,
 

#### **Padding training sequence**

In [37]:
# Step A: Find the maximum sequence length
len_list = []
for sequence in training_sequence:
    len_list.append(len(sequence))

max_len = max(len_list)

# Step B: Apply padding to all sequences
padded_sequence = []
for sequence in training_sequence:
    # Generate the required number of zeros and add them to the front
    padded_sequence.append([0] * (max_len - len(sequence)) + sequence)

# Check the final result
padded_sequence

[[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 2],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 2, 3],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 2, 3, 4],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 5, 6],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 5, 6, 7],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 5, 6, 7, 8],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 5, 6, 7, 8, 9],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 5, 6, 7, 8, 9, 10],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 5, 6, 7, 8, 9, 10, 11],
 [0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  5,
  6,
  7,
  8,
  9,
  10,
  11,
  12],
 [0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  5,
  6,
  7,
  8,
  9,
  10,
  11,
  12,
  13],
 [0,
  0,
  0,
  0,
  0,
  0,
  0,
 

# **Convert to Tensor and Covert X and y**

Before we can feed our perfectly padded sequences into a PyTorch model, we must convert our standard Python lists into **PyTorch Tensors**. After the conversion, we will apply the slicing logic we just learned to separate our inputs ($X$) from our targets ($y$).

**The Code:**

```python
# Convert the Python list into a PyTorch Tensor
padded_training_sequence = torch.tensor(padded_sequence, dtype=torch.long)

# Verify the tensor
padded_training_sequence

# Extract Features (X) and Targets (y)
X = padded_training_sequence[:, :-1]
y = padded_training_sequence[:, -1]

```

**Line-by-Line Explanation:**

* `torch.tensor(padded_sequence, ...)`: This function takes our standard Python list (`padded_sequence`) and converts it into a PyTorch tensor, which is the specialized matrix format required for deep learning operations.
* `dtype=torch.long`: This is a crucial step! It forces the numbers to be 64-bit integers. Since our data consists of vocabulary IDs (whole numbers like 1, 2, 21), PyTorch's `Embedding` layer and `CrossEntropyLoss` function strictly require the data type to be `torch.long`. If we don't specify this, PyTorch might throw a data type error during training.
* `X = padded_training_sequence[:, :-1]`: Slices the 2D tensor to select all rows and all columns *except* the very last one. These act as our input features (the context words).
* `y = padded_training_sequence[:, -1]`: Slices the 2D tensor to select all rows but *only* the very last column. This isolates the exact word our model needs to predict (the target).

---

### 🛠️ টেনসরে (Tensor) কনভার্ট করা এবং $X$ ও $y$ আলাদা করা (বাংলা ভার্সন)

আমাদের প্যাডিং করা সিকোয়েন্সগুলো PyTorch মডেলে পাঠানোর আগে, সাধারণ পাইথন লিস্টগুলোকে অবশ্যই **PyTorch Tensor**-এ রূপান্তর করতে হবে। কনভার্ট করার পর, আমরা আমাদের শেখা স্লাইসিং লজিক ব্যবহার করে ইনপুট ($X$) এবং আউটপুট ($y$) আলাদা করে ফেলব।

**কোড:**

```python
# পাইথন লিস্টকে PyTorch টেনসরে কনভার্ট করা
padded_training_sequence = torch.tensor(padded_sequence, dtype=torch.long)

# টেনসরটি চেক করা
padded_training_sequence

# ফিচার (X) এবং টার্গেট (y) আলাদা করা
X = padded_training_sequence[:, :-1]
y = padded_training_sequence[:, -1]

```

**কোডের লাইন-বাই-লাইন এক্সপ্ল্যানেশন:**

* `torch.tensor(padded_sequence, ...)`: এই ফাংশনটি আমাদের তৈরি করা সাধারণ পাইথন লিস্টটিকে একটি PyTorch টেনসরে রূপান্তর করে, যা ডিপ লার্নিংয়ের ক্যালকুলেশনের জন্য বিশেষভাবে তৈরি একটি ম্যাট্রিক্স ফরম্যাট।
* `dtype=torch.long`: এটি খুবই গুরুত্বপূর্ণ একটি অংশ! এটি ডেটা টাইপকে ৬৪-বিট ইন্টিজারে (পূর্ণসংখ্যা) সেট করে দেয়। যেহেতু আমাদের ডেটাগুলো মূলত বিভিন্ন শব্দের ID বা নাম্বার (যেমন: ১, ২, ২১), তাই PyTorch-এর `Embedding` লেয়ার এবং লস ক্যালকুলেশনের ফাংশন (`CrossEntropyLoss`) ঠিকমতো কাজ করার জন্য এই `torch.long` ডেটা টাইপটি ব্যবহার করা বাধ্যতামূলক। এটি না দিলে ট্রেনিংয়ের সময় এরর (error) আসতে পারে।
* `X = padded_training_sequence[:, :-1]`: টেনসরটিকে স্লাইস করে সবগুলো সারি এবং একদম শেষ কলামটি বাদে বাকি সব কলাম আলাদা করে নেয়। এগুলো হলো আমাদের ইনপুট বা ফিচার ($X$)।
* `y = padded_training_sequence[:, -1]`: টেনসরটিকে স্লাইস করে সবগুলো সারি থেকে শুধুমাত্র একদম শেষের কলামটি আলাদা করে নেয়। এটি হলো আমাদের টার্গেট ($y$), যা মডেলকে প্রেডিক্ট করতে হবে।

#### **Raw code Learnig purpose**

In [38]:
padded_training_sequence = torch.tensor(padded_sequence, dtype=torch.long)
padded_training_sequence

tensor([[  0,   0,   0,  ...,   0,   1,   2],
        [  0,   0,   0,  ...,   1,   2,   3],
        [  0,   0,   0,  ...,   2,   3,   4],
        ...,
        [  0,   0, 158,  ...,   3,  40,   1],
        [  0, 158, 223,  ...,  40,   1,  41],
        [158, 223,  10,  ...,   1,  41,  21]])

In [39]:
X = padded_training_sequence[:,:-1]

In [40]:
X

tensor([[  0,   0,   0,  ...,   0,   0,   1],
        [  0,   0,   0,  ...,   0,   1,   2],
        [  0,   0,   0,  ...,   1,   2,   3],
        ...,
        [  0,   0, 158,  ...,   2,   3,  40],
        [  0, 158, 223,  ...,   3,  40,   1],
        [158, 223,  10,  ...,  40,   1,  41]])

In [41]:
y = padded_training_sequence[:,-1]

In [42]:
y

tensor([  2,   3,   4,   6,   7,   8,   9,  10,  11,  12,  13,   1,  14,  15,
         16,  17,   6,  18,  19,  20,  21,  16,  22,  23,  24,  25,  26,   6,
         27,  28,  29,  30,   5,  31,  32,  21,  33,  34,  35,  36,   1,  37,
         38,   1,  39,   3,  40,   1,  41,  42,  43,   1,  44,  21,  45,   1,
         46,  47,  48,  49,  50,  51,  52,   1,   3,  21,  53,  54,  55,  56,
          6,  57,  58,  50,  59,  37,  60,  61,  21,   1,   3,  10,  63,  64,
         65,  66,  67,  68,  40,  24,  69,  21,  71,   6,  72,  73,   1,  74,
         50,  75,  48,  20,  76,  77,  48,  78,  21,   1,   3,  79,  80,  81,
         10,   1,  82,  64,  42,  83,  84,  85,  86,  40,  87,  21,  89,  90,
         91,  61,  92,  93,  94,  95,  96,   1,  97,  21,   1,  99, 100, 101,
        102, 103, 104, 101,  14, 105,  56,   6, 106, 107,  21,  27, 109, 110,
        111,  61,   1,  74,  38, 112, 113,  26, 114, 115,  21, 116,   6, 117,
        118, 119, 120, 121, 122, 123,   6,  18, 124,  21, 126, 1

#### **Convert to Tensor and Covert X and y**

In [43]:
# Convert the Python list into a PyTorch Tensor
padded_training_sequence = torch.tensor(padded_sequence, dtype=torch.long)

# Verify the tensor
padded_training_sequence

# Extract Features (X) and Targets (y)
X = padded_training_sequence[:, :-1]
y = padded_training_sequence[:, -1]

In [44]:
X

tensor([[  0,   0,   0,  ...,   0,   0,   1],
        [  0,   0,   0,  ...,   0,   1,   2],
        [  0,   0,   0,  ...,   1,   2,   3],
        ...,
        [  0,   0, 158,  ...,   2,   3,  40],
        [  0, 158, 223,  ...,   3,  40,   1],
        [158, 223,  10,  ...,  40,   1,  41]])

In [45]:
y

tensor([  2,   3,   4,   6,   7,   8,   9,  10,  11,  12,  13,   1,  14,  15,
         16,  17,   6,  18,  19,  20,  21,  16,  22,  23,  24,  25,  26,   6,
         27,  28,  29,  30,   5,  31,  32,  21,  33,  34,  35,  36,   1,  37,
         38,   1,  39,   3,  40,   1,  41,  42,  43,   1,  44,  21,  45,   1,
         46,  47,  48,  49,  50,  51,  52,   1,   3,  21,  53,  54,  55,  56,
          6,  57,  58,  50,  59,  37,  60,  61,  21,   1,   3,  10,  63,  64,
         65,  66,  67,  68,  40,  24,  69,  21,  71,   6,  72,  73,   1,  74,
         50,  75,  48,  20,  76,  77,  48,  78,  21,   1,   3,  79,  80,  81,
         10,   1,  82,  64,  42,  83,  84,  85,  86,  40,  87,  21,  89,  90,
         91,  61,  92,  93,  94,  95,  96,   1,  97,  21,   1,  99, 100, 101,
        102, 103, 104, 101,  14, 105,  56,   6, 106, 107,  21,  27, 109, 110,
        111,  61,   1,  74,  38, 112, 113,  26, 114, 115,  21, 116,   6, 117,
        118, 119, 120, 121, 122, 123,   6,  18, 124,  21, 126, 1

In [46]:
len(X)

555

In [47]:
len(y)

555

# **Custom Dataset and Dataloader Class**


In PyTorch, preparing data for the training loop requires two essential components:

1. **Dataset:** A class that stores all the data and knows how to retrieve a single item (one sequence and its target) at a time.
2. **DataLoader:** A wrapper around the Dataset that efficiently groups the data into "batches" (e.g., 32 sequences at once) and shuffles them so the model doesn't just memorize the order of the story.

**The Code:**

```python
# Define the Custom Dataset class
class CustomDataset(Dataset):
    def __init__(self, X, y):
        self.X = X
        self.y = y

    def __len__(self):
        return len(self.X)

    def __getitem__(self, idx):
        return self.X[idx], self.y[idx]
    
# Instantiate the Dataset
dataset = CustomDataset(X, y)

# Create the DataLoader
dataloader = DataLoader(dataset, batch_size=32, shuffle=True)

# Test the DataLoader
for input, output in dataloader:
    print(input, output)

```

**Line-by-Line Explanation:**

* `class CustomDataset(Dataset):`: We create a custom class that inherits from PyTorch's built-in `Dataset` class. This tells PyTorch that this class will act as our official data container.
* `def __init__(self, X, y):`: The constructor method. It receives our features ($X$) and targets ($y$) and stores them inside the class.
* `def __len__(self):`: PyTorch requires this method. It simply returns the total number of samples (rows) in our dataset so the DataLoader knows exactly how much data there is.
* `def __getitem__(self, idx):`: Another required method. When PyTorch asks for data at a specific index (`idx`), this function fetches the corresponding feature row (`self.X[idx]`) and its target word (`self.y[idx]`).
* `dataset = CustomDataset(X, y)`: We create an actual object of our custom class, passing in our sliced tensors.
* `dataloader = DataLoader(..., batch_size=32, shuffle=True)`: The DataLoader takes our `dataset` and organizes it. `batch_size=32` means it will serve 32 sequences to the model at a time. `shuffle=True` randomizes the order of the sequences in every epoch, which is critical for the model to learn the actual grammar/patterns instead of just memorizing the story line by line.
* `for input, output in dataloader:`: A test loop that grabs one batch of data at a time. Here, `input` will be a matrix of $32 \times (\text{max\_len}-1)$, and `output` will be a list of 32 target words.

---

### 🛠️ স্টেপ ৬: কাস্টম ডেটাসেট এবং ডেটালোডার (বাংলা ভার্সন)

PyTorch-এ মডেল ট্রেনিংয়ের জন্য ডেটাকে প্রস্তুত করতে দুটি গুরুত্বপূর্ণ টুল ব্যবহার করতে হয়:
১. **Dataset:** এটি এমন একটি ক্লাস যা সমস্ত ডেটা নিজের কাছে জমা রাখে এবং জানে কীভাবে একটি নির্দিষ্ট ইনডেক্স থেকে একটি ফিচার এবং তার টার্গেট বের করে আনতে হয়।
২. **DataLoader:** এটি Dataset-এর ওপর একটি মোড়ক (wrapper) হিসেবে কাজ করে। এর কাজ হলো ডেটাকে ছোট ছোট "ব্যাচ" বা ভাগে (যেমন: একসাথে ৩২টি সিকোয়েন্স) ভাগ করা এবং সেগুলোকে এলোমেলো (shuffle) করে মডেলের কাছে পাঠানো, যাতে মডেল গল্পের লাইনগুলো মুখস্থ না করে লজিক শিখতে পারে।

**কোড:**

```python
# Custom Dataset ক্লাস তৈরি করা
class CustomDataset(Dataset):
    def __init__(self, X, y):
        self.X = X
        self.y = y

    def __len__(self):
        return len(self.X)

    def __getitem__(self, idx):
        return self.X[idx], self.y[idx]
    
# Dataset-এর অবজেক্ট তৈরি করা
dataset = CustomDataset(X, y)

# DataLoader তৈরি করা
dataloader = DataLoader(dataset, batch_size=32, shuffle=True)

# DataLoader টেস্ট করা
for input, output in dataloader:
    print(input, output)

```

**কোডের লাইন-বাই-লাইন এক্সপ্ল্যানেশন:**

* `class CustomDataset(Dataset):`: আমরা PyTorch-এর বিল্ট-ইন `Dataset` ক্লাসকে ইনহেরিট (inherit) করে একটি কাস্টম ক্লাস তৈরি করেছি। এর ফলে PyTorch বুঝতে পারে যে এই ক্লাসটি আমাদের ডেটা কন্টেইনার হিসেবে কাজ করবে।
* `def __init__(self, X, y):`: এটি ইনিশিয়ালাইজার বা কনস্ট্রাক্টর মেথড। এটি আমাদের ফিচার ($X$) এবং টার্গেট ($y$) গ্রহণ করে এবং ক্লাসের ভেতরে সেভ করে রাখে।
* `def __len__(self):`: PyTorch-এর জন্য এই মেথডটি থাকা বাধ্যতামূলক। এটি মূলত আমাদের ডেটাসেটে মোট কতগুলো স্যাম্পল বা ডেটা আছে, সেই সংখ্যাটি (length) রিটার্ন করে।
* `def __getitem__(self, idx):`: এটিও একটি বাধ্যতামূলক মেথড। PyTorch যখন কোনো নির্দিষ্ট ইনডেক্সের (`idx`) ডেটা চায়, তখন এই ফাংশনটি ঠিক সেই ইনডেক্সের ফিচার (`self.X[idx]`) এবং টার্গেট (`self.y[idx]`) রিটার্ন করে।
* `dataset = CustomDataset(X, y)`: আমরা আমাদের টেনসরগুলো ($X$ এবং $y$) দিয়ে কাস্টম ক্লাসের একটি অবজেক্ট বা ইনস্ট্যান্স তৈরি করেছি।
* `dataloader = DataLoader(..., batch_size=32, shuffle=True)`: DataLoader আমাদের `dataset`-কে প্রসেস করে। `batch_size=32`-এর মানে হলো মডেল ট্রেনিংয়ের সময় এটি একসাথে ৩২টি করে সিকোয়েন্স পাঠাবে। আর `shuffle=True`-এর মানে হলো এটি প্রতিবার ডেটার সিরিয়াল এলোমেলো করে দেবে, যা মডেলের ভালো পারফরম্যান্সের জন্য খুবই জরুরি।
* `for input, output in dataloader:`: এটি একটি টেস্ট লুপ যা DataLoader থেকে একবারে এক ব্যাচ ডেটা নিয়ে আসে। এখানে `input` হবে ৩২টি ফিচারের একটি ম্যাট্রিক্স এবং `output` হবে ৩২টি টার্গেট শব্দের একটি লিস্ট।

In [48]:
class CustomDataset(Dataset):
  def __init__(self, X, y):
    self.X = X
    self.y = y

  def __len__(self):
    return len(self.X)

  def __getitem__(self, idx):
    return self.X[idx], self.y[idx]

In [49]:
dataset = CustomDataset(X, y)

In [50]:
len(dataset)

555

In [51]:
dataloader = DataLoader(dataset, batch_size=32, shuffle=True)

In [52]:
for input, output in dataloader:
  print(input, output)

tensor([[  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
           0,   0,   0,   0,   0,   6,  33,  34,  35,  36],
        [  0,   0,   0,   0,   0,   0,   0,   0,   0, 110, 175, 176, 177,  10,
         178, 179,  10,  50, 180, 181,  98,   1, 169, 182],
        [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
           0,  98,   1,  99, 100, 101, 102, 103, 104, 101],
        [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
           0,   0,   0,   0,   1,   7,  37,  10, 272, 253],
        [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0, 108,  27, 109, 110,
         111,  61,   1,  74,  38, 112, 113,  26, 114, 115],
        [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
           0,   0,   0,   0,   0,   0,   0, 125, 126, 127],
        [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,  11,
         184,  76,  92,   1,   3, 185,  36, 186, 187,  50],
        [  0,   0,   0,   0

# **Shape of X and Y before LSTM Cell**


### 📐 Step 7: Tracking Tensor Shapes & The Embedding Layer (English Version)

Before we feed our data into the LSTM cell, it is crucial to track the shapes (dimensions) of our tensors.

#### 1. Initial Shapes of $X$ and $y$

Let's assume our dataset resulted in 555 sequences, and after padding, the maximum sequence length is 24.

* **`X.shape` $\rightarrow$ `torch.Size([555, 24])**`: This means we have 555 rows (sentences/sequences), and each row contains 24 features (words/indices).
* **`y.shape` $\rightarrow$ `torch.Size([555])**`: This means we have 555 rows, but only 1 target value (the next word to predict) for each row.

#### 2. The Embedding Transformation

An LSTM cell cannot understand simple integers (like `[1, 2, 21]`). It needs dense vectors. Therefore, we pass $X$ through an **Embedding Layer**.

Let's assume we want to represent each word using a vector of 100 numbers (Embedding Dimension = 100). When our 2D tensor $X$ passes through this layer, it gets expanded into a 3D tensor:

```text
       X (Input Tensor)
     Shape: (555, 24)
            │
            ▼
    Embedding Layer
 (100 numbers per word)
            │
            ▼
 Embedded Output Tensor
   Shape: (555, 24, 100)
            │
            ▼
        LSTM Cell

```

**Breaking down the new shape `(555, 24, 100)`:**

* **`555` (Batch Size):** Total number of sentences/sequences.
* **`24` (Sequence Length):** The number of words in each sentence.
* **`100` (Input Size/Embedding Dim):** The number of features (numbers) used to represent each individual word.

This final 3D tensor is exactly what the LSTM cell expects as its input!

---

### 📐 স্টেপ ৭: টেনসরের শেইপ ট্র্যাক করা এবং এমবেডিং লেয়ার (বাংলা ভার্সন)

LSTM সেলে ডেটা পাঠানোর আগে আমাদের টেনসরের শেইপ বা ডাইমেনশন ট্র্যাক করা অত্যন্ত জরুরি।

#### ১. $X$ এবং $y$-এর প্রাথমিক শেইপ

ধরে নিই, আমাদের ডেটাসেট থেকে ৫৫৫টি সিকোয়েন্স তৈরি হয়েছে এবং প্যাডিং করার পর প্রতিটি সিকোয়েন্সের সাইজ বা দৈর্ঘ্য হলো ২৪।

* **`X.shape` $\rightarrow$ `torch.Size([555, 24])**`: এর মানে হলো আমাদের ৫৫৫টি সারি (row বা বাক্য) আছে এবং প্রতিটি সারিতে ২৪টি ফিচার (শব্দ বা ইনডেক্স) আছে।
* **`y.shape` $\rightarrow$ `torch.Size([555])**`: এর মানে হলো আমাদের ৫৫৫টি সারি আছে, কিন্তু প্রতিটি সারির জন্য আউটপুট বা টার্গেট ভ্যালু মাত্র ১টি (যেই শব্দটি প্রেডিক্ট করতে হবে)।

#### ২. এমবেডিং ট্রান্সফরমেশন (Embedding Transformation)

LSTM সেল সাধারণ ইন্টিজার বা পূর্ণসংখ্যা (যেমন `[১, ২, ২১]`) বুঝতে পারে না; এর জন্য ডেন্স ভেক্টর (Dense Vector) প্রয়োজন। তাই, আমরা $X$-কে একটি **Embedding Layer**-এর ভেতর দিয়ে পাঠাব।

ধরে নিই, আমরা প্রতিটি শব্দকে ১০০টি নাম্বারের একটি ভেক্টর দিয়ে উপস্থাপন করতে চাই (Embedding Dimension = 100)। যখন আমাদের 2D টেনসর $X$ এই লেয়ারের ভেতর দিয়ে যাবে, তখন এটি একটি 3D টেনসরে রূপান্তরিত হবে:

```text
       X (ইনপুট টেনসর)
     শেইপ: (555, 24)
            │
            ▼
     Embedding Layer
  (প্রতি শব্দের জন্য ১০০ নাম্বার)
            │
            ▼
  Embedded আউটপুট টেনসর
   শেইপ: (555, 24, 100)
            │
            ▼
        LSTM Cell

```

**নতুন শেইপ `(555, 24, 100)`-এর বিস্তারিত অর্থ:**

* **`555` (Batch Size):** মোট বাক্যের বা সিকোয়েন্সের সংখ্যা।
* **`24` (Sequence Length):** প্রতিটি বাক্যে থাকা মোট শব্দের সংখ্যা।
* **`100` (Input Size/Embedding Dim):** প্রতিটি শব্দকে বোঝানোর জন্য ব্যবহৃত নাম্বারের পরিমাণ (ভেক্টর সাইজ)।

এই 3D টেনসরটিই হলো পারফেক্ট ইনপুট, যা আমরা সরাসরি আমাদের মডেলের LSTM সেলে পাঠাব!

In [54]:
X.shape

torch.Size([555, 24])

In [55]:
y.shape

torch.Size([555])